packages/rxfire/firestore/collection/index.ts

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { firestore } from 'firebase/app';
+import { fromCollectionRef } from '../fromRef';
+import { Observable } from 'rxjs';
+import { map, filter, scan } from 'rxjs/operators';
+
+const ALL_EVENTS: firestore.DocumentChangeType[] = [
+  'added',
+  'modified',
+  'removed'
+];
+
+/**
+ * Create an operator that determines if a the stream of document changes
+ * are specified by the event filter. If the document change type is not
+ * in specified events array, it will not be emitted.
+ */
+const filterEvents = (events?: firestore.DocumentChangeType[]) =>
+  filter((changes: firestore.DocumentChange[]) => {
+    let hasChange = false;
+    changes.forEach(change => {
+      if (events.indexOf(change.type) >= 0) {
+        hasChange = true;
+      }
+    });
+    return hasChange;
+  });
+
+/**
+ * Create an operator that filters out empty changes. We provide the
+ * ability to filter on events, which means all changes can be filtered out.
+ * This creates an empty array and would be incorrect to emit.
+ */
+const filterEmpty = filter(
+  (changes: firestore.DocumentChange[]) => changes.length > 0
+);
+
+/**
+ * Creates a new sorted array from a new change.
+ * @param combined
+ * @param change
+ */
+function combineChange<T>(
+  combined: firestore.DocumentChange[],
+  change: firestore.DocumentChange
+): firestore.DocumentChange[] {
+  switch (change.type) {
+    case 'added':
+      if (
+        combined[change.newIndex] &&
+        combined[change.newIndex].doc.id == change.doc.id
+      ) {
+        // Not sure why the duplicates are getting fired
+      } else {
+        combined.splice(change.newIndex, 0, change);
+      }
+      break;
+    case 'modified':
+      // When an item changes position we first remove it
+      // and then add it's new position
+      if (change.oldIndex !== change.newIndex) {
+        combined.splice(change.oldIndex, 1);
+        combined.splice(change.newIndex, 0, change);
+      } else {
+        combined.splice(change.newIndex, 1, change);
```

Comment:
This is just a more difficult way of saying:

```typescript
combined[change.newIndex] = change;
``` 

Can we swap it out and add a comment?

Reply:
fixed